### PR TITLE
Test guidelines updated to write proper tests and avoid duplicates

### LINKED ISSUE #3687: Update test guidelines and link it to the pr template
/kind feature

## Which functionality do you think we should add?

Add a field  in the pr template that will document the process of how to avoid duplicate test.

## Why is this needed?

It helps contributor to get a quick reference of our test guidelines. It will provide benefits like: know if the pr has any duplicate test script or not and many more.

### PR DESCRIPTION
**Pull request by prietyc123**
**What type of PR is this?**

/kind documentation

**What does does this PR do / why we need it**:

Proper test guidelines has been updated and also linked to the PR template

**Which issue(s) this PR fixes**:

Fixes #3687 

**PR acceptance criteria**:

- [ ] Unit test 

- [ ] Integration test 

- [x] Documentation 

**How to test changes / Special notes to the reviewer**:

### CODE REVIEW

File: docs/dev/test-architecture.adoc [hunk 1]
```diff
@@ -184,7 +184,13 @@ Context("when running help for storage command", func() {
 ----
 +
 
-* Due to parallel test run support make sure that the should run in isolation, otherwise the test result will lead to race condition. To achieve this ginkgo provides some in build functions `BeforeEach`, `AfterEach` etc.
+* Don't create a new test spec for the steps which can be run with the existing specs. Also try to avoid adding the functionality which are part of another feature test file, only add if it is a key requirement.
++ 
+
+* Spec level conditions, pre and post requirements should be run in ginkgo in build functions `JustBeforeEach` and `JustAfterEach`
```
Comment (amitkrout) [inline docs/dev/test-architecture.adoc:190]: ```in build functions `JustBeforeEach` and `JustAfterEach```` -> ```built-in tear down steps `JustBeforeEach` and `JustAfterEach````

Comment (prietyc123) [reply docs/dev/test-architecture.adoc]: Done

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 2]
```diff
@@ -30,4 +30,6 @@ Fixes #?
 
 - [ ] Documentation 
 
+- [ ] I have read the [test guidelines](https://github.com/openshift/odo/blob/master/docs/dev/test-architecture.adoc) properly
```
Comment (amitkrout) [inline .github/PULL_REQUEST_TEMPLATE.md:33]: The sentence sense ungentle to me. This can be done through a polite statement something like  
```
Read the [test guidelines](https://github.com/openshift/odo/blob/master/docs/dev/test-architecture.adoc)
```

Comment (prietyc123) [reply .github/PULL_REQUEST_TEMPLATE.md:33]: @Team Need your attention because this change going to affect all while creating a pr (Specifically test contribution)

File: docs/dev/test-architecture.adoc [hunk 3]
```diff
@@ -184,7 +184,13 @@ Context("when running help for storage command", func() {
 ----
 +
 
-* Due to parallel test run support make sure that the should run in isolation, otherwise the test result will lead to race condition. To achieve this ginkgo provides some in build functions `BeforeEach`, `AfterEach` etc.
+* Don't create a new test spec for the steps which can be run with the existing specs. Also try to avoid adding the functionality which are part of another feature test file, only add if it is a key requirement.
```
Comment (cdrage) [inline docs/dev/test-architecture.adoc:187]: Sentences are a bit weird..

This may be better:

**Do not create a new [Spec](https://onsi.github.io/ginkgo/#structuring-your-specs) if there is already an existing one.**

Instead of what you current have. Both sentences say the same thing (you're just repeating the first sentence in the second sentence).

File: docs/dev/test-architecture.adoc [hunk 4]
```diff
@@ -184,7 +184,13 @@ Context("when running help for storage command", func() {
...
++ 
+
+* Spec level conditions, pre and post requirements should be run in ginkgo built-in tear down steps `JustBeforeEach` and `JustAfterEach`
```
Comment (cdrage) [inline docs/dev/test-architecture.adoc:190]: This is confusing.. maybe link to Ginkgo documentation? Even I don't understand this much :(

File: docs/dev/test-architecture.adoc [hunk 5]
```diff
@@ -184,7 +184,13 @@ Context("when running help for storage command", func() {
...
++
+
+* Due to parallel test run support make sure that the spec should run in isolation, otherwise the test result will lead to race condition. To achieve this ginkgo provides some in build functions `BeforeEach`, `AfterEach` etc.
```
Comment (cdrage) [inline docs/dev/test-architecture.adoc:193]: Link to Ginkgo documentation. Don't use `etc.`

You can shorten this to:

* Some tests are unable to be run in parallel. You can achieve this by running XX (link to Ginkgo doc or example) with your test.

Or something like that.

File: docs/dev/test-architecture.adoc [hunk 6]
```diff
@@ -238,10 +244,16 @@ Expect(output).To(ContainSubstring("No file changes detected, skipping build"))
 ----
 +
 
-* If oc, odo or generic library you are looking for is not present in helper package then create a new library function as per the scenario requirement.
+* If oc, odo or generic library you are looking for is not present in helper package then create a new library function as per the scenario requirement.
```
Comment (cdrage) [inline docs/dev/test-architecture.adoc:247]: I think this can actually be removed. Sort of self explanatory to create a new library function if it's not available...

File: docs/dev/test-architecture.adoc [hunk 7]
```diff
@@ -238,10 +244,16 @@ Expect(output).To(ContainSubstring("No file changes detected, skipping build"))
 ----
 +
 
-* If oc, odo or generic library you are looking for is not present in helper package then create a new library function as per the scenario requirement.
+* If oc, odo or generic library you are looking for is not present in helper package then create a new library function as per the scenario requirement. 
++
+
+* Avoid logic implementation within the test files as it unnecessarily overloads the tests. Always create a new function with all the required logics within the helper package. Check for code duplicacy within helper and test files.
```
Comment (cdrage) [inline docs/dev/test-architecture.adoc:250]: Avoid logic implementation? Bit confused what you mean..

Comment (cdrage) [reply docs/dev/test-architecture.adoc:250]: Can remove part about `Check for code duplicacy within helper and test files.` that's pretty self-explanatory

Comment (cdrage) [reply docs/dev/test-architecture.adoc:250]: Maybe we can shorten it to: Avoid unnecessary function implementation within test files. Check to see if there is a helper function already implement

?

File: docs/dev/test-architecture.adoc [hunk 8]
```diff
@@ -238,10 +244,16 @@ Expect(output).To(ContainSubstring("No file changes detected, skipping build"))
...
++
+
+* If you are looking for delay with a specific feature test, don't use `time.Sleep()` function, Instead use `timeout`.
```
Comment (cdrage) [inline docs/dev/test-architecture.adoc:253]: Ironically, we use this within our code
```sh
  1 tests/helper/helper_run.go|41 col 2| time.Sleep(timeout)                                                                                                                                                                                                                                                                                                                 
  2 tests/helper/helper_run.go|92 col 4| time.Sleep(time.Duration(intervalSeconds) * time.Second)
  3 || 
  4 tests/helper/helper_http.go|34 col 4| time.Sleep(time.Duration(interval) * time.Second)
  5 tests/helper/helper_http.go|45 col 3| time.Sleep(time.Duration(interval) * time.Second)
  6 || 
  7 tests/helper/helper_generic.go|140 col 3| time.Sleep(timeout)
  8 || 
  9 tests/integration/generic_test.go|349 col 4| time.Sleep(1 * time.Second)
```

But yeah, good tip, we shouldn't ever have time.sleep in the tests.

File: docs/dev/test-architecture.adoc (hunk 4 above)
Comment (prietyc123) [reply docs/dev/test-architecture.adoc:190]: Yes, adding ginkgo documentation seems hard and fast requirement but IMO we should add ginkgo link to the very first line of test guidelines. TBH adding the link for every steps doesn't make sense to me.

File: docs/dev/test-architecture.adoc (hunk 5 above)
Comment (prietyc123) [reply docs/dev/test-architecture.adoc:193]: > Some tests are unable to be run in parallel.

But I would say stating out the reason best explain the use of `BeforeEach`, `AfterEach` and so on. For adding link I would again say the same as in https://github.com/openshift/odo/pull/3726#discussion_r468378476

File: docs/dev/test-architecture.adoc (hunk 6 above)
Comment (prietyc123) [reply docs/dev/test-architecture.adoc:247]: Instead of just removing this I would kind of more inclined towards combining https://github.com/openshift/odo/pull/3726/files#diff-27813ce26d5cb87e1c303086adb6db24R247-R250

File: docs/dev/test-architecture.adoc (hunk 7 above)
Comment (prietyc123) [reply docs/dev/test-architecture.adoc:250]: > Maybe we can shorten it to: Avoid unnecessary function implementation within test files. Check to see if there is a helper function already implement

+1, In addition to this I will also add library function implementation in this point.

File: docs/dev/test-architecture.adoc [hunk 9]
```diff
@@ -138,7 +138,7 @@ Refer to the odo clean test link:https://github.com/openshift/odo/blob/master/te
 
 *Test guidelines:*
 
-Please follow certain protocol before contributing to odo tests. This helps in how to contribute in link:https://github.com/openshift/odo/tree/master/tests[`odo tests`].
+Please follow certain protocol before contributing to odo tests. This helps in how to contribute in link:https://github.com/openshift/odo/tree/master/tests[`odo tests`]. For better understanding of writing test please refer ginkgo link:https://onsi.github.io/ginkgo/#getting-ginkgo[documentation].
```
Comment (amitkrout) [inline docs/dev/test-architecture.adoc:141]: ... please refer Ginkgo link:https://onsi.github.io/ginkgo/#getting-ginkgo[documentation] and Ginkgo's preferred matcher library Gomega link:http://onsi.github.io/gomega/[documentation]

Comment (prietyc123) [reply docs/dev/test-architecture.adoc]: Done

File: docs/dev/test-architecture.adoc (hunk 4 above)
Comment (amitkrout) [reply docs/dev/test-architecture.adoc:190]: @prietyc123 No need to give the full doc link, you should provide the link to the teardown feature just make the test guideline intutive. For example

Spec level conditions, pre and post requirements should be run in ginkgo built-in tear down  link:http://onsi.github.io/ginkgo/#structuring-your-specs[steps] `JustBeforeEach` and `JustAfterEach`

File: docs/dev/test-architecture.adoc (hunk 8 above)
Comment (amitkrout) [reply docs/dev/test-architecture.adoc:250]: > If you are looking for delay with a specific feature test, don't use `time.Sleep()` function, Instead use `timeout`.

@prietyc123 I am making the statement intuitive. PTAL 

If you are looking for delay with a specific feature test, don't use hard `time.Sleep()` function. Yes, you can use but as a polling interval of maximum duration. Check the helper package[link] for more such reference 

For example: 
```
func RetryInterval(maxRetry, intervalSeconds int, program string, args ...string) string {
	for i := 0; i < maxRetry; i++ {
		session := CmdRunner(program, args...)
		session.Wait()
		if session.ExitCode() == 0 {
			time.Sleep(time.Duration(intervalSeconds) * time.Second)
		} else {
			Consistently(session).ShouldNot(gexec.Exit(0), runningCmd(session.Command))
			return string(session.Err.Contents())
		}
	}
	Fail(fmt.Sprintf("Failed after %d retries", maxRetry))
	return ""
}
```
Or in Ginkgo there is a in build timeout feature [link] that you can make use of it.

File: docs/dev/test-architecture.adoc (hunk 4 above)
Comment (prietyc123) [reply docs/dev/test-architecture.adoc:190]: @amitkrout But its already there https://github.com/openshift/odo/pull/3726/files#diff-27813ce26d5cb87e1c303086adb6db24R141 before jotting down the key points. Does it make sense of referring same doc everytime. AFAIU users would anyway get the idea from the referred doc before going through further points.

File: docs/dev/test-architecture.adoc (hunk 5 above)
Comment (amitkrout) [reply docs/dev/test-architecture.adoc:193]: You can combine ```* Spec level conditions, pre and post requirements should be run in ginkgo built-in tear down steps `JustBeforeEach` and `JustAfterEach```` and ```* Due to parallel test run support make sure that the spec should run in isolation, otherwise the test result...``` like 

You can combine ```* Spec level conditions, pre and post requirements should be run in ginkgo built-in tear down steps `JustBeforeEach` and `JustAfterEach. To make spec isolation for parallel run you can use and update the same teardown steps as per the requirement.

File: docs/dev/test-architecture.adoc (hunk 4 above)
Comment (amitkrout) [reply docs/dev/test-architecture.adoc:190]: @prietyc123 https://onsi.github.io/ginkgo/#getting-ginkgo is the whole ginkgo doc and http://onsi.github.io/ginkgo/#structuring-your-specs[steps] is a quick reference of that particular steps. These teardown steps are used more frequently in the test suite, so for a contribute it will be like a first look reference instead of going through the whole doc.

File: tests/template/template_cleantest_test.go [hunk 10]
```diff
@@ -35,98 +35,53 @@ var _ = Describe("Example of a clean test", func() {
...
-				//helper.CmdShouldPass("odo component create nodejs")
-				helper.CmdShouldPass("odo", "component", "create", "nodejs")
-				//TODO: verify that config was properly created
-				helper.CmdShouldPass("odo", "push")
-				//TODO: verify resources on cluster
-			})
-
+		// go back to original project after each test
+		JustAfterEach(func() {
+			oc.SwitchProject(originalProject)
```
Comment (amitkrout) [inline tests/template/template_cleantest_test.go:64]: Now it is no where used in the test suite. It has been replaced with ```--project``` flag. So can remove it completely. And try to replicate the template as per our current test suite use.

File: docs/dev/test-architecture.adoc (hunk 4 above)
Comment (amitkrout) [reply docs/dev/test-architecture.adoc:190]: Check for your reference - https://github.com/openshift/odo/pull/3726/files#diff-27813ce26d5cb87e1c303086adb6db24R270.

File: tests/template/template_cleantest_test.go (hunk 10 above)
Comment (prietyc123) [reply tests/template/template_cleantest_test.go:64]: I can see we are still using this here - https://github.com/openshift/odo/blob/master/tests/integration/component.go#L67-L75 and some more places also. So do you mean we should remove from those places also or how is it

Comment (prietyc123) [reply tests/template/template_cleantest_test.go]: @amitkrout Yes, I don't feel any significance of `oc.SwitchProject(originalProject)` in our test script. So I removed it from template but as we are still using it at some places lets remove this from there also.
